refactor: add context parameters to inner functions

paralin opened the pull request:
Add context parameters to eliminate many instances of context.TODO and improve code consistency.

Drop a flaky test (increment).

cc @dennwc 

By the way: this is part of a larger refactor of cayley at https://github.com/aperturerobotics/cayley - any plans to continue work on cayley? Anything I've done in that branch that seems useful @dennwc ? thanks for your hard work on this.

paralin committed:
refactor: add context parameters to inner functions

Add context parameters to eliminate many instances of context.TODO and improve code consistency.

Fix flaky test on btree (increment) by marking it as NoLocks and NoTx.

Signed-off-by: Christian Stewart <christian@aperture.us>

dennwc reviewed (changes requested):
Thank you for taking the time to do it! :bow: Most changes make total sense, but two raise concern.

As for Cayley, some changes seem destructive and I'm not sure what was the reasoning behind them. In general, I have not plans to continue working on Cayley anymore.

Having said that, Cayley was in a frozen state for a while now, and we failed to find a maintainer that can keep it at least updated. So if you can keep it in a good shape and/or continue the development - that would be great.

paralin commented on kv/helpers.go:
```diff
@@ -53,10 +53,3 @@ func Each(ctx context.Context, tx Tx, fnc func(k Key, v Value) error, opts ...It
 }
-
-// CreateBucket is a helper to create buckets upfront without writing any key-value pairs to it.
-func CreateBucket(_ context.Context, tx Tx, key Key) error {
```
@dennwc The reason was because there's no need to create keys with nil values (buckets). Removing this as well as the version number (constant at 2) reduced the number of read/writes significantly in my use case. 

Will drop it for this PR since it's not suitable for your upstream repo.

paralin commented on kv/kvtest/kvtest.go:
```diff
@@ -75,7 +74,6 @@ var testList = []struct {
 	{name: "basic", test: basic},
 	{name: "ro", test: readonly},
 	{name: "seek", test: seek},
-	{name: "increment", test: increment, txOnly: true, concurrent: true},
```
@dennwc It wasn't passing about 25% of the time against the btree backend (in-memory). I figured that the test itself must be implemented incorrectly since surely the in-memory btree has transaction semantics. But now upon further review I see that it does not (it's marked as not safe for concurrency use)

I'll add the test back for this PR but will submit another one to fix the test. I think it might make sense to just fix btree to be concurrency safe as opposed to changing the test to not run on btree.

paralin replied:
Fixed it by adding the NoLocks and NoTx options for btree, missed those before!

paralin commented on tuple/kv/upgrade.go:
```diff
@@ -532,20 +532,20 @@ func (tbl *tupleTable) scan(f *tuple.Filter) *tupleIterator {
+		// TODO: support descending order
+		return &tupleIterator{err: fmt.Errorf("descending order is not supported")}
 	}
-	// FIXME: support limit
```
👍🏽

paralin commented:
@dennwc Addressed your comments, thanks for the review! I think I can fix the "FIXME" limit parameter in a later PR.

My more destructive / overhaul changes to cayley and hidalgo are because I'm working on simplifying + cleaning up those two to use as libraries for an embedded graph DB store. For debugging purposes it's much easier to read the following output with ascii keys:

```
Get("v/CAm2dbq2XZbxcee36m325UhBCHU") => data(1) found(true) err(<nil>)
Get("m/horizon") => data(8) found(true) err(<nil>)
ScanPrefix("sp/1") => callback("sp/1:3", len(1))
ScanPrefix("sp/1") => callback("sp/1:9", len(2))
Get("l/4") => data(8) found(true) err(<nil>)
```

The changes to add contexts everywhere and get rid of context.TODO has taken me a couple of days, but I'm just about done with it now. This fixed a number of critical bugs that I ran into where sometimes an iterator would fail with context.Canceled, and return zero results instead of the error.

Since I'm running Cayley within another transaction system, the transactional semantics not working quite right in Cayley is not a problem at all.

I really like the structure of the code as a library (the API w/ gizmo and graphql is impressive as well!).

The design of the Path and morphisms is very clever and I'm still working to wrap my head around it all.

dennwc commented:
Thanks a lot!
